src/librustc_trans/trans/intrinsic.rs

```diff
@@ -589,217 +582,86 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
+                Some((width, signed)) =>
+                    match c {
+                        "ctlz" => count_zeros_intrinsic(bcx, &format!("llvm.ctlz.i{}", width),
+                                                        llargs[0], call_debug_location),
+                        "cttz" => count_zeros_intrinsic(bcx, &format!("llvm.cttz.i{}", width),
+                                                        llargs[0], call_debug_location),
+                        "ctpop" => Call(bcx, ccx.get_intrinsic(&format!("llvm.ctpop.i{}", width)),
+                                        &llargs, None, call_debug_location),
+                        "bswap" => {
+                            if width == 8 {
+                                llargs[0] // byte swap a u8/i8 is just a no-op
+                            } else {
+                                Call(bcx, ccx.get_intrinsic(&format!("llvm.bswap.i{}", width)),
+                                        &llargs, None, call_debug_location)
+                            }
+                        }
+                        "add_with_overflow" =>
+                            if signed {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.sadd.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            } else {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.uadd.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            },
+                        "sub_with_overflow" =>
+                            if signed {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.ssub.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            } else {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.usub.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            },
+                        "mul_with_overflow" =>
+                            if signed {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.smul.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            } else {
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.umul.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            },
```

Comment:
All of these could use the same implementation, with "s" or "u" decided by `signed` and the operation name from `name[..3]`.